nucleus/utils.py

```diff
@@ -422,3 +434,87 @@ def paginate_generator(
             yield json_value
         if not next_token:
             break
+
+def fetch_image(s3_url: str) -> Image.Image:
+    s3_bucket = s3_url.split("//")[-1].split("/")[0]
+    s3_output = s3_url.split("//")[-1].split(s3_bucket)[-1][1:]
+    image = Image.open(
+        boto3.resource("s3").Bucket(s3_bucket).Object(s3_output).get()["Body"]
+    )
+    return image
+
+def fetch_chip(s3_prefix: str) -> tuple[str, str]:
+    s3_bucket = s3_prefix.split("//")[-1].split("/")[0]
+    s3_output = s3_prefix.split("//")[-1].split(s3_bucket)[-1][1:]
+    try:
+        boto3.resource("s3").Bucket(s3_bucket).Object(s3_output + '.jpeg').load()
+    except:
+        return None, None
+    try:
+        boto3.resource("s3").Bucket(s3_bucket).Object(s3_output + '.json').load()
+    except:
+        return s3_prefix + '.jpeg', None
+    return s3_prefix + '.jpeg', s3_prefix + '.json'
+
+def upload_chip(s3_prefix: str, chipped_image: Image.Image, chipped_annotations: [dict]) -> tuple[str, str]:
+    s3_bucket = s3_prefix.split("//")[-1].split("/")[0]
+    s3_output = s3_prefix.split("//")[-1].split(s3_bucket)[-1][1:]
+    byteio = io.BytesIO()
+    chipped_image.save(byteio, format="jpeg")
+    byteio.seek(0)
+    try:
+        boto3.resource("s3").Bucket(s3_bucket).Object(s3_output + '.jpeg').upload_fileobj(byteio)
+    except Exception as err:
+        print("hi", err)
+        return None, None
+    if len(chipped_annotations) == 0:
+        return s3_prefix + '.jpeg', None
+    try:
+        boto3.resource("s3").Bucket(s3_bucket).Object(s3_output + '.json').put(Body=json.dumps(chipped_annotations))
+    except Exception as err:
+        print("sup", err)
+        return None, None
+    return s3_prefix + '.jpeg', s3_prefix + '.json'
+
+def generate_offsets(w: int, h: int, chip_size: int, stride_size: int):
+    xs = np.arange(0, w - stride_size, chip_size - stride_size)
+    ys = np.arange(0, h - stride_size, chip_size - stride_size)
+    if len(xs) > 1:
```

Comment:
I like this feature (shrink the stride_size so the last chip ends at the end of the image instead of extending beyond it) but will everyone? It should probably be mentioned in the docstring at least. You could also make it optional but then you have to decide if you want to drop the last (partial) chip or pad it with something.